allennlp/data/semparse/type_declarations/type_declaration.py

```diff
@@ -254,3 +257,128 @@ def make_VariableExpression(self, name):
+
+
+def _substitute_placeholder_type(_type: Type, basic_type: BasicType) -> Type:
+    """
+    Takes a type with placeholders, and a basic type and substitutes all occurrences of the placeholder with
+    that type.
+    """
+    # TODO (pradeep): This assumes there's just one placeholder in the type. So this doesn't work with
+    # ``reverse`` yet, which has two placeholders.
+    if _type == ANY_TYPE:
+        return basic_type
+    if isinstance(_type, BasicType):
+        return _type
+    return ComplexType(_substitute_placeholder_type(_type.first, basic_type),
+                       _substitute_placeholder_type(_type.second, basic_type))
+
+
+def _make_production_string(source: Type, target: Union[List[Type], Type]) -> str:
+    return "%s -> %s" % (str(source), str(target))
+
+
+def _get_complex_type_productions(complex_type: ComplexType) -> List[Tuple[str, str]]:
+    """
+    Takes a complex type without any placeholders and returns all productions that lead to it, starting
+    from the most basic return type. For example, if the complex is `<a,<<b,c>,d>>`, this gives the
+    following tuples
+    ('<<b,c>,d>', '<<b,c>,d> -> [<a,<<b,c>,d>>, a]')
+    ('d', 'd -> [<<b,c>,d>, <b,c>]')
+    """
+    all_productions = []
+    while isinstance(complex_type, ComplexType) and not complex_type == ANY_TYPE:
+        all_productions.append((str(complex_type.second), _make_production_string(complex_type.second,
+                                                                                  [complex_type,
+                                                                                   complex_type.first])))
+        complex_type = complex_type.second
+    return all_productions
+
+
+def get_valid_actions(name_mapping: Dict[str, str],
+                      type_signatures: Dict[str, Type],
+                      basic_types: Set[Type]) -> Dict[str, Set[str]]:
+    """
+    Generates all the valid actions starting from each non-terminal. For terminals of a specific type, we
+    simply add to valid actions, productions from thee types to the terminals. Among those types, we keep
+    track of all the non-basic types (i.e., function types). For those types, we infer the list of productions
+    that start from a basic type leading to them.
+    For complex types that do not contain ANY_TYPE or placeholder
+    types, this is straight-forward. For example, if the complex type is <e,<r,<d,r>>>, the productions should
+    be [r -> [<d,r>, r], <d,r> -> [<r,<d,r>>, r], <r,<d,r>> -> [<e,<r,<d,r>>>, e]].
+    However, if the complex type contains ANY_TYPE, we substitute it with all possible basic types. If it
+    contains placeholders, the substituyions need to be constrained. For example, for <#1,#1>, <e,r> is not a
+    valid substitution.
+    """
+    valid_actions: Dict[str, Set[str]] = defaultdict(set)
+
+    complex_types = set()
+    for name, alias in name_mapping.items():
+        if not alias in type_signatures:
```

Comment:
It's more idiomatic to do `if alias not in type_signatures:`

Reply:
Also, when would this happen, where you have a name mapping that doesn't show up in `type_signatures`?

Reply:
This happens only when the `name` is `lambda`. Changed the conditional to simply check for that.